maven-core/src/main/java/org/apache/maven/lifecycle/internal/MojoDescriptorCreator.java

```diff
@@ -225,10 +225,10 @@ else if ( numTokens == 3 )
             plugin = findPluginForPrefix( prefix, session );
         }
 
-        int executionIdx = goal.indexOf( '@' );
+        int executionIdx = goal.toString().indexOf( '@' );
         if ( executionIdx > 0 )
         {
-            goal = goal.substring( 0, executionIdx );
+            goal = new StringBuilder(goal.substring(0, executionIdx));
```

Comment:
That looks like the kind of operation we'd like to avoid : a conversion to String, and a conversion back to a StringBuilder.  Use `goal.setLength( executionIdx )` ?
Not sure of the overall benefit if we manipulate `StringBuilder` to avoid a few string manipulation.   Maybe the use of the `StringBuilder` could be limited to refactoring the following code:
```
             goal = tok.nextToken();

             // This won't be valid, but it constructs something easy to read in the error message
             while ( tok.hasMoreTokens() )
             {
                 goal += ":" + tok.nextToken();
             }
```